### PR TITLE
Snapshotting and sync could cause a dead lock TranslogWriter

### PR DESCRIPTION
**Pull request by bleskes**
#18360 introduced an extra lock in order to allow writes while syncing the translog. This caused a potential deadlock (see [1])  with snapshotting code where we first acquire the instance lock, followed by a sync (which acquires the syncLock). However, the sync logic acquires the syncLock first, followed by the instance lock.

I considered solving this by not syncing the translog on snapshot - I think we can get away with just flushing it. That however will create subtleties around snapshoting and whether operations in them are persisted. I opted instead to have slightly uglier code with nest synchronized, where the scope of the change is contained to the TranslogWriter class alone. If people feel differently, I can take the other approach.

[1]

```
  2> "elasticsearch[node_s1][index][T#1]" ID=8144 BLOCKED on org.elasticsearch.index.translog.TranslogWriter@7527aa29 owned by "elasticsearch[node_s1][generic][T#2]" ID=8079
  2>    at org.elasticsearch.index.translog.TranslogWriter.syncUpTo(TranslogWriter.java:225)
  2>    - blocked on org.elasticsearch.index.translog.TranslogWriter@7527aa29
  2>    - locked java.lang.Object@46936c8a

  2> "elasticsearch[node_s1][generic][T#2]" ID=8079 BLOCKED on java.lang.Object@46936c8a owned by "elasticsearch[node_s1][index][T#1]" ID=8144
  2>    at org.elasticsearch.index.translog.TranslogWriter.syncUpTo(TranslogWriter.java:219)
  2>    - blocked on java.lang.Object@46936c8a
  2>    at org.elasticsearch.index.translog.TranslogWriter.sync(TranslogWriter.java:151)
  2>    at org.elasticsearch.index.translog.TranslogWriter.newSnapshot(TranslogWriter.java:201)
  2>    - locked org.elasticsearch.index.translog.TranslogWriter@7527aa29
```

### CODE REVIEW

**Comment by bleskes** [comment]
@s1monw can you take a look?

File: core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java [hunk 1]
```diff
@@ -19,6 +19,7 @@
 
 package org.elasticsearch.index.translog;
 
+import com.carrotsearch.randomizedtesting.annotations.Repeat;
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: This should be removed. :)

Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java]: aye - missing a checkbox in Intellij :D

**Comment by jasontedor** [comment]
LGTM.

**Pull request merged**